parsley-debug/shared/src/main/scala/parsley/debugger/internal/DebugContext.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2020 Parsley Contributors <https://github.com/j-mie6/Parsley/graphs/contributors>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+package parsley.debugger.internal
+
+import scala.collection.mutable.ListBuffer
+
+import org.typelevel.scalaccompat.annotation.unused
+import parsley.debugger.ParseAttempt
+import parsley.debugger.combinator.defaultRules
+import parsley.internal.deepembedding.frontend.LazyParsley
+
+// Class used to hold details about a parser being debugged.
+// This is normally held as a value inside an implicit variable.
+// Anything caught by the toStringRules will have a parse result of that type toString-ed for memory
+// efficiency.
+private [parsley] class DebugContext(val toStringRules: Seq[Any => Boolean] = defaultRules) {
+    // Create a new dummy root of the tree that will act as filler for the rest of the tree to build
+    // off of (as there is no "nil" representation for the tree... other than null, which should be
+    // avoided in Scala wherever possible).
+    private val dummyRoot: TransientDebugTree =
+        TransientDebugTree("ROOT", "ROOT", "NIL")
+
+    // Context's checkStack vanished.
+    // Migrating it here for futureproofing.
+    private var checkStack: ListBuffer[(Int, Int, Int)] =
+        ListBuffer()
```

Comment:
```scala
private var checkStack = mutable.ListBuffer.empty[(Int, Int, Int)]
```

Reply:
on second thought, why is it a `var` of a mutable collection?

Reply:
> on second thought, why is it a `var` of a mutable collection?

Not going to lie, I forgot `.clear()` exists when writing that part of the code, whoops